### ISSUE TITLE
[User] Implement Streaming LLM - Make the inference more efficient

### ISSUE THREAD

**Issue by errorsandwarnings** [ggerganov/llama.cpp#3440]
# Prerequisites
Context length limit is an issue on all LLMs. The following repository and associated paper is demonstrating that keeping the 4 initial tokens will enable a infinite context length on most common LLMs without sacrificing performance or efficiency.

Code : https://github.com/mit-han-lab/streaming-llm

Paper reference inside the repo which demonstrates the attention-sink effect of LLMs and how to take advantage of it.

# Current Behavior

There is a limit on context length defined mostly by pre-training. Other approaches like rope or sliding window have their pros and cons, none of them can get to a higher context length than this apporach.

**Comment by ggerganov** [comment]
Yup, I've already read the paper and the good news is that after #3228 this technique is trivial to support in `llama.cpp`.
It's technically already implemented by setting `n_keep == 4` in `main`

**Comment by nathanodle** [comment]
> They also demonstrate how memory efficient this technique is (without it, it runs OOM very early). Does that basically mean infinite context with steady and low VRAM usage? That would be revolutionary and render vectorDBs obsolete.

This isn't infinite context the way you're reading it.  It uses a novel modification to attention so that you can generate "infinite" output

**Comment by staviq** [comment]
If somebody can confirm if I'm understanding that paper right, I'd be grateful:

They are proposing a solution for infinite *text* length, not infinite *context* length, right ? And their observation is that the consistency of "internal state" depends heavily on first tokens, so naively keeping initial tokens and implementing sliding context window on the rest, allows the LLM to keep its sanity intact, and first tokens are "important" because appending tokens from one side, makes the first token "exist" for N iterations, second token for N-1 iterations etc, so the first tokens are "seen" by all subsequent iterations but not the other way around ?

**Comment by errorsandwarnings** [comment]
> > They also demonstrate how memory efficient this technique is (without it, it runs OOM very early). Does that basically mean infinite context with steady and low VRAM usage? That would be revolutionary and render vectorDBs obsolete.
> 
> This isn't infinite context the way you're reading it. It uses a novel modification to attention so that you can generate "infinite" output

Yes, it took me a while to understand and their newly posted FAQ clarifies it. I have edited the title. I believe it is a good technique to be part of this project.

**Comment by Dampfinchen** [comment]
True, this one however...

https://github.com/tomaarsen/attention_sinks

is potentially relevant. It's about the input sequence length, so the context this time. Look at that steady memory usage!
![272347418-3a4c5634-cc1b-42d1-a35a-afb376a4f970](https://github.com/ggerganov/llama.cpp/assets/59751859/1ea2e6cf-dfbd-4c41-b6d9-cf5d79aeab87)

**Comment by KerfuffleV2** [comment]
> It's about the input sequence length, so the context this time. Look at that steady memory usage!

The key point is the input sequence length can be very long but the context the model is considering stays constant. So you could feed it a book, have it write a book worth of content but it won't "remember" or take into account what was in the sequence 4,096 tokens ago or whatever.

**Comment by errorsandwarnings** [comment]
The VRAM usage is impressive.

**Comment by phillip-kravtsov** [comment]
I think we need `n_keep=4` but also `n_discard=1` here to correctly implement StreamingLLM: https://github.com/ggerganov/llama.cpp/blob/master/examples/main/main.cpp#L504. 

Assuming that the llama.cpp shifting of the k-cache matches the behavior described in the paper, where the keys are rotated according to their position in the cache rather than the text (which I believe they are but not certain)

**Comment by ggerganov** [comment]
> I think we need n_keep=4 but also n_discard=1 here to correctly implement StreamingLLM:

That's correct, although I don't think one really needs to shift each new token. The benefit would be marginal.

**Comment by ggerganov** [comment]
There is no longer expensive re-evaluation in `main` and `server` since #3228 was merged. Instead of re-evaluating, we are now "shifting" the KV cache which is a relatively cheap operation and in some sense is equivalent to the approach proposed in the paper. I would say we even have an advantage because we have to the option to set `n_discard == 8` for example which would make RoPE recalculation 1 every 8 tokens instead of on each token as it is done in StreamingLLM

**Comment by Tostino** [comment]
Just wondering, does this give us the option to choose where the sliding window begins? e.g. I have a prompt template as seen here:
```
<#meta#>
- Date: 2023-10-05
- Task: chat
<#system#>
You are a conversational AI having a turn based chat with a user.
<#chat#>
<#user#>
Message 1
<#bot#>
Response 1
<#user#>
Message 2
<#user_context#>
Some Context
<#bot#>
Response 2
<#user#>
Message 3
<#bot#>
Response 3
```

Could I anchor this portion:
```
<#meta#>
- Date: 2023-10-05
- Task: chat
<#system#>
You are a conversational AI having a turn based chat with a user.
<#chat#>
```
And have the kv cache shift only over the chat portion?

Or am I just misunderstanding things?

**Comment by ggerganov** [comment]
Yes - count the tokens in that portion and set `n_keep` equal to that number

**Comment by Tostino** [comment]
And is n_keep configurable during inference time? One of the features I was planning on, was integrating an ensemble LLM which can modify the prompt template at specific points during inference. E.g. to change the current task, or change the system prompt to align with the current problem that is being worked on in the response, and then resume inference for example. So the number of tokens in that window may change.

**Comment by ggerganov** [comment]
You can modify it - the API is very flexible. Though to achieve your goal, it would take more than just updating `n_keep`. One way is to have separate context for each prompt that you evaluate once at the start. Or you can have one context and different sequence ids for the different prompts.

**Comment by MoonRide303** [comment]
@ggerganov Is it possible to use StreamingLLM / Windowed Attention with Attention Sinks method as defined in paper via main, currently? If so - how? Something like `--keep 4` or `--keep 8` will be enough? I don't see `--n_keep` or `--n-discard` in main options list. Results in paper looked really impressive - basically infite chats with low PPL (they've tested it up to 4 million token), without extra computational / memory costs - shouldn't it be default setting for compatible models? Or I am missing something, and there are some downsides not mentioned in the paper?

**Comment by ggerganov** [comment]
The discussion above is still valid, you can use the `--keep` argument to control the sink size:

https://github.com/ggerganov/llama.cpp/blob/4399f13fb9462cd06f3f154d0aee738425000fea/common/common.cpp#L1386

The `n_discard` is currently not exposed for modification through the CLI, but it is easy to adjust in the code:

https://github.com/ggerganov/llama.cpp/blob/4399f13fb9462cd06f3f154d0aee738425000fea/examples/main/main.cpp#L554

As discussed earlier, it is likely required to set `n_discard == 1` in order to match the implementation from the paper.

> Results in paper looked really impressive ..  Or I am missing something, and there are some downsides not mentioned in the paper?

You might be missing that even though the PPL is low, it does no mean that the entire information from the processed context is "visible". The model will still "see" just the last `n_ctx` tokens. More info: 

- hypothesis: https://github.com/ggerganov/llama.cpp/pull/2268#issuecomment-1784158879
- test: https://github.com/ggerganov/llama.cpp/pull/3856
- conclusion: https://github.com/ggerganov/llama.cpp/pull/2268#issuecomment-1784754822

**Comment by reversTeam** [comment]
Related: #21105 adds `llama_set_attn_mask()` which could be used to implement explicit attention sink control — setting a custom mask that always allows attending to the initial tokens regardless of window position. The per-head variant allows different heads to have different sink behaviors.
